### PR TITLE
Make the device id on e2e key upload optional

### PR DESCRIPTION
**Pull request by richvdh**
We should now be able to get our device_id from the access_token, so the device_id on the upload request is optional. Where it is supplied, we should check that it matches.

Also make sure that the device_id is registered in the devices table, for both new e2e keys and those already in the database.

Finally, delete e2e keys from the database when we delete a device.

### CODE REVIEW

**Comment by richvdh** [comment]
matrixbot: retest this please

File: synapse/rest/client/v2_alpha/keys.py [hunk 1]
```diff
@@ -51,23 +54,47 @@ class KeyUploadServlet(RestServlet):
       },
     }
     """
-    PATTERNS = client_v2_patterns("/keys/upload/(?P<device_id>[^/]*)", releases=())
+    PATTERNS = client_v2_patterns("/keys/upload(/(?P<device_id>[^/]+))?$",
+                                  releases=(), v2_alpha=False)
```
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/keys.py]: Might want to update sytest to use unstable rather than v2_alpha?

Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/keys.py]: I see that you already have.

**Comment by NegativeMjark** [comment]
matrixbot: retest this please

File: synapse/rest/client/v2_alpha/keys.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/keys.py]: Is there a reason to remove the v2_alpha path?

File: synapse/rest/client/v2_alpha/keys.py [hunk 2]
```diff
@@ -51,23 +54,47 @@ class KeyUploadServlet(RestServlet):
...
         user_id = requester.user.to_string()
-        # TODO: Check that the device_id matches that in the authentication
-        # or derive the device_id from the authentication instead.
 
         body = parse_json_object_from_request(request)
 
+        if device_id is not None:
+            # passing the device_id here is deprecated; however, we allow it
+            # for now for compatibility with older clients. But if a device_id
+            # was given here and in the auth, they must match.
```
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/keys.py]: Why do they need to match? Do they match in vector now, or will vector need to be updated for this to work?

File: synapse/rest/client/v2_alpha/keys.py (hunk 1 above)
Comment (richvdh) [inline synapse/rest/client/v2_alpha/keys.py]: just that it's kinda BS, and Vector is using `unstable`.

File: synapse/rest/client/v2_alpha/keys.py (hunk 2 above)
Comment (richvdh) [inline synapse/rest/client/v2_alpha/keys.py]: It seems to be silly to allow you to upload keys for a different device.

And sadly this does need fixing in vector. I'm trying to fix vector, but that is making me cry too.

File: synapse/rest/client/v2_alpha/keys.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/keys.py]: Is Vector the only thing using that API? I though Tor had got a client using the e2e stuff.

File: synapse/rest/client/v2_alpha/keys.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/keys.py]: Might it be worth holding off adding this check until vector is updated? Or are we concluding that it's okay to break it on the old vector?

File: synapse/rest/client/v2_alpha/keys.py (hunk 1 above)
Comment (richvdh) [inline synapse/rest/client/v2_alpha/keys.py]: yeah he does. https://github.com/torhve/weechat-matrix-protocol-script/blob/master/matrix.lua

and I see he's using v2_alpha

FUCK IT

this is why we shouldn't publish apis on v2_alpha until they are stable.

File: synapse/rest/client/v2_alpha/keys.py (hunk 2 above)
Comment (richvdh) [inline synapse/rest/client/v2_alpha/keys.py]: yeah and it's gonna break Tor's stuff too. argh fuckery I hate my life.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**